syntax = "proto3";

service RaftService {
  rpc RequestVote(TermIdMessage) returns (TermResultMessage);
  rpc AppendEntries(TermIdMessage) returns (TermResultMessage);
}

service ClientService {
  rpc Connect(Empty) returns (TermIdMessage);
  rpc GetLeader(Empty) returns (IdAddressMessage);
  rpc Suspend(IntMessage) returns (Empty);
  rpc SetVal(SetValMessage) returns (SetValResponseMessage);
  rpc GetVal(GetValMessage) returns (GetValResponseMessage);
}

message Empty {}

message TermIdMessage {
  int32 term = 1;
  int32 id = 2;
}

message TermResultMessage {
  int32 term = 1;
  bool result = 2;
}

message IdAddressMessage{
  int32 id = 1;
  string address = 2;
}

message IntMessage {
  int32 value = 1;
}

message SetValMessage {
  string key = 1;
  string value = 2;
}

message GetValMessage {
  string key = 1;
}

message SetValResponseMessage {
  bool success = 1;
}

message GetValResponseMessage {
  bool success = 1;
  string value = 2;
}